### PR TITLE
Introduce the SchemeConfig structure as a way to configure the engine.

### PR DESCRIPTION
**Pull request by marmeladema**
Only call_stack_depth_limit is declared at this time.

### CODE REVIEW

File: engine/src/scheme.rs [hunk 1]
```diff
@@ -214,10 +247,22 @@ impl<'s> Scheme {
...
 
+    /// Retrieves the scheme configuration as immutable.
+    /// Usually used to read configuration values.
+    pub fn get_config(&self) -> &SchemeConfig {
+        &self.config
+    }
+
+    /// Retrieves the scheme configuration as mutable.
+    /// Usually used to write configuration values.
+    pub fn get_mut_config(&mut self) -> &mut SchemeConfig {
```
Comment (inikulin) [inline engine/src/scheme.rs]: The common notation is to add `mut` at the end, so it should be `get_config_mut`

File: engine/src/scheme.rs [hunk 2]
```diff
@@ -183,6 +183,37 @@ impl<'i> Display for ParseError<'i> {
...
+            fn default() -> Self {
+                Self::new()
+            }
+        }
+    };
+}
+
+declare_config!(
+    #[doc="Call stack depth limit"]
+    let call_stack_depth_limit: usize = std::usize::MAX
```
Comment (inikulin) [inline engine/src/scheme.rs]: In this particular case usage of macro seems to be an overkill: everything can be easily and clearly expressed using standard language syntax. Declare fields in structure, set default values in `Default` trait implementation.

File: engine/src/lex.rs [hunk 3]
```diff
@@ -68,6 +68,12 @@ pub enum LexErrorKind {
         #[cause]
         mismatch: TypeMismatchError,
     },
+
+    #[fail(
+        display = "call stack depth should not exceed {} levels, but {} levels",
```
Comment (inikulin) [inline engine/src/lex.rs]: The sentence is unfinished `,but {} levels ... ?`

**Comment by marmeladema** [comment]
I managed to implement this check on top of the JSON serialized AST instead. I don't have a strong opinion whether or not this should be part of the core Wirefilter?

**Comment by RReverser** [comment]
> I managed to implement this check on top of the JSON serialized AST instead.

Personally, I'd keep it there for now, and when we have enough of them, introduce a generic validation mechanism based on the known usecases.

**Comment by marmeladema** [comment]
Then lets close this for now :+1:

**Pull request closed without merging**